File: ErrorStateKalmanFilter/QuasiStationaryNorthFinding/Functions/state_transition.m
function [Phi] = state_transition(F,Ts,model_order)
% Input Parameters
% F  -> System State Space Model
% Ts -> Sampling Periode
% model_order -> Taylor Series Truncation Order
% Output Parameters
% Phi -> State Transition Matrix

dim = size(F,1);

switch model_order
    case 1
        Phi = (eye(dim,dim)) + (F*Ts);        
    case 2
        Phi = (eye(dim,dim)) + (F*Ts) + (0.5*F^2*Ts^2);    
    case 3
        Phi = (eye(dim,dim)) + (F*Ts) + (0.5*F^2*Ts^2) + (1/6*F^3*Ts^3);
    % case 4
    % Phi = eye(dim,dim) + F*Ts + 0.5*F^2*Ts^2 + 1/6*F^3*Ts^3 + 1/24*F^4*Ts^4;
    % case 5
    % Phi = eye(dim,dim) + F*Ts + 0.5*F^2*Ts^2 + 1/6*F^3*Ts^3 + 1/24*F^4*Ts^4 + 1/120*F^5*Ts^5;
end

end

File: ErrorStateKalmanFilter/ESKFcpp/testCpp/StateTransition.cpp
#include <iostream>
#include <Eigen/Dense>

using namespace Eigen;

MatrixXd state_transition(const MatrixXd &f_mat, double Ts, int model_order) {
    int dim = f_mat.rows();
    MatrixXd Phi = MatrixXd::Identity(dim, dim);

    switch (model_order) {
        case 1:
            Phi += (f_mat * Ts);
            break;
        case 2:
            Phi += (f_mat * Ts) + (0.5 * f_mat * f_mat * Ts * Ts);
            break;
        case 3:
            Phi += (f_mat * Ts) + (0.5 * f_mat * f_mat * Ts * Ts) + ((1.0 / 6.0) * f_mat * f_mat * f_mat * Ts * Ts * Ts);
            break;
        // Add more cases for higher orders as needed
        // case 4:
        //     Phi += (f_mat * Ts) + (0.5 * f_mat * f_mat * Ts * Ts) + ((1.0 / 6.0) * f_mat * f_mat * f_mat * Ts * Ts * Ts) + ((1.0 / 24.0) * f_mat * f_mat * f_mat * f_mat * Ts * Ts * Ts * Ts);
        //     break;
        // case 5:
        //     Phi += (f_mat * Ts) + (0.5 * f_mat * f_mat * Ts * Ts) + ((1.0 / 6.0) * f_mat * f_mat * f_mat * Ts * Ts * Ts) + ((1.0 / 24.0) * f_mat * f_mat * f_mat * f_mat * Ts * Ts * Ts * Ts) + ((1.0 / 120.0) * f_mat * f_mat * f_mat * f_mat * f_mat * Ts * Ts * Ts * Ts * Ts);
        //     break;
        default:
            std::cerr << "Invalid model order." << std::endl;
            break;
    }

    return Phi;
}

int main() {

    MatrixXd F(3, 3);
    F << 1, 2, 3,
         4, 5, 6,
         7, 8, 9;

    double Ts = 0.1;
    int modelOrder = 2;

    MatrixXd result = state_transition(F, Ts, modelOrder);

    std::cout << "State Transition Matrix:\n" << result << std::endl;

    return 0;
}
// g++ -I C:/toolbox/eigen-3.4.0/ -o StateTransition StateTransition.cpp